src/misc/cache/class/lv_cache_sc_da.c

```diff
@@ -0,0 +1,507 @@
+                      void * user_data)
+{
+    LV_UNUSED(user_data);
+
+    lv_cache_sc_da_t * da = (lv_cache_sc_da_t *)cache;
+
+    LV_ASSERT_NULL(da);
+    LV_ASSERT_NULL(entry);
+
+    if(da == NULL || entry == NULL) {
+        return;
+    }
+
+    const size_t entry_size = lv_cache_entry_get_size(da->cache.node_size);
+    uint8_t * da_entry_to_remove = (uint8_t *)lv_cache_entry_get_data(entry);
+
+    void * last_da_entry;
+    lv_cache_entry_t * last_cache_entry;
+    get_entry(da, cache->size - 1, &last_da_entry, &last_cache_entry);
+
+    if(da_entry_to_remove != last_da_entry) {
+        lv_memcpy(da_entry_to_remove, last_da_entry, entry_size);
+    }
+    cache->size -= 1;
+}
+
+static void drop_cb(lv_cache_t * cache, const void * key, void * user_data)
+{
+    lv_cache_sc_da_t * da = (lv_cache_sc_da_t *)cache;
+
+    LV_ASSERT_NULL(da);
+    LV_ASSERT_NULL(key);
+
+    if(da == NULL || key == NULL) {
+        return;
+    }
+    lv_cache_entry_t * entry = cache->clz->get_cb(cache, key, user_data);
+    if(!entry) {
+        return;
+    }
+
+    const size_t entry_size = lv_cache_entry_get_size(da->cache.node_size);
+    uint8_t * da_entry_to_remove = (uint8_t *)lv_cache_entry_get_data(entry);
+
+    void * last_da_entry;
+    lv_cache_entry_t * last_cache_entry;
+    get_entry(da, cache->size - 1, &last_da_entry, &last_cache_entry);
+
+    if(da_entry_to_remove != last_da_entry) {
+        lv_memcpy(da_entry_to_remove, last_da_entry, entry_size);
+    }
+    cache->ops.free_cb(da_entry_to_remove, user_data);
+    cache->size -= 1;
+}
+
+static void drop_all_cb(lv_cache_t * cache, void * user_data)
+{
+    lv_cache_sc_da_t * da = (lv_cache_sc_da_t *)cache;
+
+    LV_ASSERT_NULL(da);
+
+    if(da == NULL) {
+        return;
+    }
+    uint32_t used_cnt = 0;
+    for(size_t i = 0; i < cache->size; ++i) {
+        void * da_entry;
+        lv_cache_entry_t * cache_entry;
+        get_entry(da, i, &da_entry, &cache_entry);
+        const int32_t refs = lv_cache_entry_get_ref(cache_entry);
+
+        if(refs > 0) {
+            LV_LOG_WARN(
+                "entry (%zu) is still referenced (%" LV_PRId32
+                ")",
+                i, refs);
+            used_cnt++;
+            continue;
+        }
+        da->cache.ops.free_cb(da_entry, user_data);
+    }
+    if(used_cnt > 0) {
+        LV_LOG_WARN("%" LV_PRId32 " entries are still referenced",
+                    used_cnt);
+    }
+
+    cache->size = 0;
+}
+
+static lv_cache_entry_t * get_possible_victim(lv_cache_sc_da_t * da, size_t index)
+{
+    LV_ASSERT(index < da->cache.size);
+
+    void * da_entry;
+    lv_cache_entry_t * cache_entry;
+    get_entry(da, index, &da_entry, &cache_entry);
+
+    const uint8_t sec_chance = get_second_chance(cache_entry);
+    const int32_t refs = lv_cache_entry_get_ref(cache_entry);
+
+    if(sec_chance == 0 && refs == 0) {
+        return cache_entry;
+    }
+    if(sec_chance == 0 && refs > 0) {
```

Comment:
Isn't it normal that an entry is being used while adding something to the cache? E.g. a thread draws an image, while and other thread wants to draw an other image. 

If it's true it should be rather an info log.